plugins/text-editor-resources/src/components/extension/link.ts

```diff
@@ -39,5 +45,156 @@ export const LinkShortcutsExtension = Extension.create<any>({
+
+interface ReferenceProps {
+  id: Ref<Doc>
+  objectclass: Ref<Class<Doc>>
+  label: string
+}
+
+export function ResolveReferenceUrlsPlugin (editor: Editor): Plugin<LinkToReferencePluginState> {
+  return new Plugin<LinkToReferencePluginState>({
+    key: resolveReferencePluginKey,
+
+    appendTransaction: (transactions, oldState, newState) => {
+      if (transactions[0]?.getMeta('linkToReference') === undefined) return undefined
+      if (editor.schema.nodes.reference === undefined) return
+
+      const references = resolveReferencePluginKey.getState(newState)?.references ?? new Map()
+
+      const { tr } = newState
+      tr.doc.descendants((node, pos) => {
+        if (!node.isText || !node.marks.some((m) => m.type.name === 'link')) return
+
+        const url = node.textContent
+        const mapping = references.get(url)
+        if (mapping === undefined) return
+
+        const replacementNode = editor.schema.nodes.reference.create(mapping)
+        const mpos = tr.mapping.map(pos)
+        tr.replaceWith(mpos, mpos + node.nodeSize, replacementNode)
+      })
+
+      if (tr.steps.length > 0) return tr
+    },
+
+    state: {
+      init () {
+        return {
+          references: new Map(),
+          queue: new Set()
+        }
+      },
+      apply (tr, prev, oldState, newState) {
+        if (tr.getMeta('linkToReference') !== undefined) {
+          const references = tr.getMeta('linkToReference').references as LinkToReferencePluginState['references']
+          const urls = new Set(references.keys())
+          return {
+            queue: new Set(Array.from(prev.queue).filter((url) => !urls.has(url))),
+            references: new Map([...prev.references, ...references])
+          }
+        }
+
+        if (!tr.docChanged || oldState.doc.eq(newState.doc)) return prev
+
+        const urls: string[] = []
+        tr.doc.descendants((node) => {
+          if (!node.isText || !node.marks.some((m) => m.type.name === 'link')) return
+          const url = node.textContent
+
+          const hasNoMapping = prev.references.has(url) && prev.references.get(url) === undefined
+          if (prev.queue.has(url) || hasNoMapping) return
+
+          urls.push(url)
+        })
+
+        const promises = urls.map(async (url) => {
+          try {
+            return [url, await getReferenceFromUrl(url)] as const
+          } catch {
+            return [url, undefined] as const
+          }
+        })
+
+        if (promises.length > 0) {
+          void Promise.all(promises).then((references) => {
+            editor.view.dispatch(editor.state.tr.setMeta('linkToReference', { references: new Map(references) }))
+          })
+        }
+
+        return {
+          references: prev.references,
+          queue: new Set([...prev.queue, ...urls])
+        }
+      }
+    }
+  })
+}
+
+async function getReferenceFromUrl (text: string): Promise<ReferenceProps | undefined> {
```

Comment:
Probably not needed to fix right now, but I would move this logic somewhere to an utility function or reused existing (if any) to ensure we keep consistent logic. For example, we already have references extension that do similar things.

Reply:
The thing is, there's no good place to put it at the momement, and there's no existing equivalent alternative. The only thing that knows how to parse links is the Workbench svelte component. It should probably be moved somewhere else, but in the Workbench refactoring context.